### PR TITLE
Vector2: Added angleTo()

### PR DESCRIPTION
**Pull request by puxiao**
Related issue: #null

**Description**

Vector2 added `.angleTo()`, just like `Vector3.angleTo()`.

### CODE REVIEW

File: src/math/Vector2.js [hunk 1]
```diff
@@ -365,6 +367,20 @@ class Vector2 {
...
+
+		const denominator = Math.sqrt( this.lengthSq() * v.lengthSq() );
+
+		if ( denominator === 0 ) return Math.PI / 2;
+
+		const theta = this.dot( v ) / denominator;
+
+		// clamp, to handle numerical problems
+
+		return Math.acos( MathUtils.clamp( theta, - 1, 1 ) );
```
Comment (Mugen87) [inline src/math/Vector2.js:380]: You copied the code over from `Vector3` however please check if the 2D case can be implemented with less instructions: https://stackoverflow.com/questions/14066933/direct-way-of-computing-clockwise-angle-between-2-vectors

**Comment by WestLangley** [comment]
```js
const x = new THREE.Vector2( 1, 0 );
const y = new THREE.Vector2( 0, - 1 );

console.log( y.angle() ); // 4.71238898038469

console.log( x.angleTo( y ) ); // -1.5707963267948966
```

**Comment by puxiao** [comment]
> ```js
> const x = new THREE.Vector2( 1, 0 );
> const y = new THREE.Vector2( 0, - 1 );
> 
> console.log( y.angle() ); // 4.71238898038469
> 
> console.log( x.angleTo( y ) ); // -1.5707963267948966
> ```



<br>

Is this the result you want?

```
angleTo( v ) {

		let angle = Math.atan2( this.cross( v ), this.dot( v ) );
		
		if( angle < 0 ){
				angle += 2 * Math.PI;
		}
		
		return angle;

}
```

**Comment by WestLangley** [comment]
Your first implementation returned a positive result -- presumably the smallest of the two angles. That is the result I would expect in three.js.

The second implementation returns a signed result. That method likely requires a different name and proper documentation.

What is your use case for needing this method?

**Comment by puxiao** [comment]
I am studying linear algebra recently.
By chance, I found that `Vector3` has `.angleTo()` but `Vector2` does not.
Emm....
In fact, I don't have a case where this method needs to be used.

**Comment by Mugen87** [comment]
We usually do not enhance the API without a compelling use case. Just because `Vector3` has a method does not mean it is required in `Vector2` or `Vector4`.

**Comment by WestLangley** [comment]
>I don't have a case where this method needs to be used.

That is what I expected. There has been no demand for this in three.js. This is a 3D library.

Also, there are many use cases: clockwise, counter-clockwise, smallest-angle, [0-2pi], [-pi, pi], signed angle, unsigned angle.

//

One option is to do what you did initially, and add the method that returns the shortest angle, unsigned.

The other option is to do nothing since there has been no demand for it.

**Comment by puxiao** [comment]
Okay, I get it, thanks.

I will close this PR.

**Comment by jespertheend** [comment]
I'm working on controls for a vehicle that only moves on a flat plane, and `Vector2.angleTo()` would have been useful here to be honest. I was hoping to keep the direction vector of the vehicle 2d to avoid it from drifting upwards.

**Comment by mrdoob** [comment]
I think the method is so tiny that is probably worth adding it 👍

**Comment by mrdoob** [comment]
Unfortunately I can't reopen the PR...

**Comment by Mugen87** [comment]
As mentioned in https://github.com/mrdoob/three.js/pull/22209#issuecomment-889045540 the method can be implemented differently and not every implementation fits all use cases. I agree with @WestLangley that `Vector2.angleTo()` should adhere the same policy as `Vector3.angleTo()`. 

@jespertheend Would the method still helpful to you if it returns the shortest (unsinged) angle between two vectors?

**Comment by jespertheend** [comment]
Yeah I think a signed angle would probably be more useful.

I also frequently run across this case with 3d vectors:
```js
const up = new Vector3(0, 1, 0);
let angle = direction.angleTo(up);
const cross = direction.cross(up);
if (cross.z < 0) {
	angle *= -1;
}
```
From a quick search in my current (4 week) three.js project I ran across this 7 times.
In comparison, this project used `angleTo` only 10 times. So the majority of those were followed by steps to make it unsigned.
Doing it with 3d vectors needs an extra measure axis though, in this case the `up` vector.

In Renda I have these four functions:
`Vec2.angleTo()`
https://github.com/rendajs/Renda/blob/16eb03603466b27ecfebef30c619c2f752a22582/src/math/Vec2.js#L400-L405
`Vec2.clockwiseAngleTo()`
https://github.com/rendajs/Renda/blob/16eb03603466b27ecfebef30c619c2f752a22582/src/math/Vec2.js#L362-L367
`Vec3.angleTo()`
https://github.com/rendajs/Renda/blob/16eb03603466b27ecfebef30c619c2f752a22582/src/math/Vec3.js#L467-L472
`Vec3.clockwiseAngleTo()`
https://github.com/rendajs/Renda/blob/16eb03603466b27ecfebef30c619c2f752a22582/src/math/Vec3.js#L513-L521

I personally think all of these could be useful, but ultimately it's a trade-off of how much you want to include.

**Comment by Mugen87** [comment]
Considering https://github.com/mrdoob/three.js/pull/22209#issuecomment-1469890346 and your implementations, `Vector2.angleTo()` should look like so: 7e32385ca8d5944b053d6572d5f8a688f836ae90. This version was also suggested by @WestLangley  in https://github.com/mrdoob/three.js/pull/22209#issuecomment-889045540.

@mrdoob If you are fine with this commit, I'll restore the PR based on that version.

**Comment by WestLangley** [comment]
hmmm...

- I expect if we add an "unsigned" `Vector2` version, someone will then want a "signed" `Vector2` version.

- As per the docs, `Vector3.angleTo()` computes the angle "between" the two vectors. Maybe it should be named `Vector3.angleBetween()`?

There are many options for implementing this. Very little demand.

**Comment by Mugen87** [comment]
> I expect if we add an "unsigned" Vector2 version, someone will then want a "signed" Vector2 version.

Yes, this will only be a matter of time. But still I think it's okay to add `Vector2.angleTo()` since we have the same for `Vector3`.

> As per the docs, Vector3.angleTo() computes the angle "between" the two vectors. Maybe it should be named Vector3.angleBetween()?

Lately, I frequently get complains that `three.js` changes its API too often and these complains are in some sense justified. We should be more strict in that regard and only renaming things when it's absolutely necessary. TBH, I don't see this need in context if `Vector3.angleTo()`.

**Comment by mrdoob** [comment]
@Mugen87 

> @mrdoob If you are fine with this commit, I'll restore the PR based on that version.

Sounds good!

**Comment by mrdoob** [comment]
@WestLangley 

> - I expect if we add an "unsigned" `Vector2` version, someone will then want a "signed" `Vector2` version.

Let's cross that bridge when we come to it 🙂

**Pull request closed without merging**